File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = set()
        maxLen = 0
        l = 0

        for i in range(len(s)):

            while s[i] in res:
                res.remove(s[l])
                l += 1
                
            res.add(s[i])
            maxLen = max(maxLen,len(res))

        return maxLen
    

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> window = new HashSet<>();
        int l = 0;
        int maxLen = 0;
        char[] arr = s.toCharArray();
        for (char c : arr){
            while (window.contains(c)){
                window.remove(arr[l]);
                l += 1;
            }
            window.add(c);
            if (window.size() > maxLen) {maxLen = window.size();}
        }

        return maxLen;
    }
}